src/psyclone/domain/lfric/lfric_extract_driver_creator.py

```diff
@@ -263,63 +292,60 @@ def flatten_reference(self, old_reference, symbol_table,
         # Now remove '_proxy' that might have been added to a variable name,
         # to preserve the expected names from a user's point of view.
         symbol_name = proxy_name_mapping.get(signature[0], signature[0])
-        # Create the new signature, e.g. f1_proyx%data --> f1
-        signature = Signature(symbol_name, signature[1:])
-
-        # We use this string as a unique tag - it must be unique since no
-        # other tag uses a '%' in the name. So even if the flattened name
-        # (e.g. f1_data) is not unique, the tatg `f1%data` is unique, and
-        # the symbol table will then create a unique name for this symbol.
 
         if isinstance(old_reference.symbol.datatype, ArrayType):
             # Vector field. Get the index that is being accessed, and
             # don't append the '%data'
-            indx = int(old_reference.children[-1].value)
+            indx = int(old_reference.indices[-1].value)
             signature = Signature(f"{symbol_name}%{indx}")
         else:
+            # Create the new signature, e.g. f1_proyx%data --> f1
             field_type = old_reference.symbol.datatype.name
             if field_type in ["field_proxy_type", "r_solver_field_proxy_type",
                               "r_tran_field_proxy_type"]:
                 # Field proxy are accessed using '%data'. Remove this to
                 # have more familiar names for the user, and also because
                 # the plain name is used in the file written.
-                signature = Signature(signature[0])
+                signature = Signature(symbol_name)
             else:
                 # Other types need to get the member added to the name,
                 # to make unique symbols (e.g. 'op_a_proxy%ncell_3d'
                 # and 'op_a_proxy%local_stencil'
                 signature = Signature(symbol_name, signature[1:])
 
+        # We use this string as a unique tag - it must be unique since no
+        # other tag uses a '%' in the name. So even if the flattened name
+        # (e.g. f1_data) is not unique, the tag `f1%data` is unique, and
+        # the symbol table will then create a unique name for this symbol.
         signature_str = str(signature)
         try:
             symbol = symbol_table.lookup_with_tag(signature_str)
         except KeyError:
-            flattened_name = self.flatten_string(signature_str)
+            flattened_name = self._flatten_signature(signature)
             symbol = DataSymbol(flattened_name, old_reference.datatype)
             symbol_table.add(symbol, tag=signature_str)
 
-        array_member = None
         current = old_reference
+        # Get the indices from a structure array access, e.g.
+        # field%data(i) --> field(i)
         while current:
             if isinstance(current, ArrayMember):
```

Comment:
Thanks for tidying this. GH isn't letting me continue the previous conversation, hence this new comment. I'm a bit concerned that this code is implicitly assuming that we either have e.g. `a%b` or `a%b(i)` (if I've understood correctly). While that might currently be true in LFRic, it's not generally true in PSyIR, i.e. we could have e.g. `a%b(i)%c`. Does this need some checking that 'current' is of the  form as a defence against future changes in LFRic data structures?

Reply:
Good point. I've added a check for `ArrayOfStructures` and abort in this case (including a test).